near-sdk/src/collections/heap.rs

```diff
@@ -0,0 +1,464 @@
+            for x in 0..n {
+                vec.push(x);
+            }
+            let mut rng = rand::thread_rng();
+            vec.shuffle(&mut rng);
+            vec
+        }
+
+        let cases = vec![10, 20, 30, 42]; // Error(GasLimitExceeded) for sizes >= 50
+        for n in cases {
+            let mut heap: Heap<u32> = Heap::new(vec![b't']);
+
+            let items = random(n);
+            for x in &items {
+                heap.insert(x);
+            }
+            assert_eq!(heap.len(), n as u64);
+
+            let mut sorted = items.clone();
+            sorted.sort();
+
+            let actual = heap.sort_iter().collect::<Vec<u32>>();
+            heap.clear();
+            assert_eq!(actual, sorted,
+                       "Sorting {:?} failed: expected {:?} but got {:?}.", items, sorted, actual);
+        }
+    }
+
+    #[test]
+    fn test_insert() {
+        test_env::setup();
+
+        let mut heap: Heap<u8> = Heap::new(vec![b't']);
+        let key = 42u8;
+        assert!(heap.lookup(&key).is_none());
+        assert_eq!(heap.len(), 0);
+
+        heap.insert(&key);
+        assert_eq!(heap.len(), 1);
+        assert_eq!(heap.lookup(&key), Some(1));
+        heap.clear();
+    }
+
+    #[test]
+    fn test_insert_duplicate() {
+        test_env::setup();
+
+        let mut heap: Heap<u8> = Heap::new(vec![b't']);
+        let key = 42u8;
+        assert!(heap.lookup(&key).is_none());
+        assert_eq!(heap.len(), 0);
+
+        heap.insert(&key);
+        heap.insert(&key);
+        assert_eq!(heap.len(), 1);
+        assert_eq!(heap.lookup(&key), Some(1));
+        assert_eq!(heap.indices.len(), 1);
+        assert_eq!(heap.elements.len(), 1);
+
+        heap.clear();
+    }
+
+    #[test]
+    fn test_remove() {
+        test_env::setup();
+
+        let mut heap: Heap<u8> = Heap::new(vec![b't']);
+        let key = 42u8;
+        assert!(heap.lookup(&key).is_none());
+
+        heap.insert(&key);
+        heap.remove(&key);
+
+        assert!(heap.lookup(&key).is_none());
+        assert_eq!(heap.len(), 0);
+
+        heap.clear();
+    }
+
+    #[test]
+    fn test_get_max() {
+        test_env::setup();
+        let mut heap: Heap<u8> = Heap::new(vec![b't']);
+
+        for x in vec![1u8, 2u8, 3u8, 4u8, 5u8] {
+            heap.insert(&x);
+        }
+
+        assert_eq!(heap.get_max(), Some(5u8));
+        assert_eq!(heap.len(), 5);
+
+        heap.clear();
+    }
+
+    #[test]
+    fn test_remove_max() {
+        test_env::setup();
+        let mut heap: Heap<u8> = Heap::new(vec![b't']);
+
+        let vec = vec![1u8, 2u8, 3u8, 4u8, 5u8];
+        for x in vec.iter() {
+            heap.insert(&x);
+        }
+
+        let n = vec.len();
+        for (i, _) in vec.iter().enumerate() {
+            assert_eq!(heap.remove_max(),  Some((n - i) as u8));
+            assert_eq!(heap.len() as usize, n - 1 - i);
+        }
+
+        assert_eq!(heap.len(), 0);
```

Comment:
Add an extra remove_maxp to test removing max from an empty heap.

Reply:
Nice catch, thanks.